### ISSUE TITLE
undefined reference to `__cranelift_probestack'

### ISSUE THREAD

**Issue by jyn514** [bytecodealliance/cranelift#1263]
- What are the steps to reproduce the issue? Can you include a CLIF test case,
  ideally reduced with the `bugpoint` clif-util command?

```clif
function u0:0() -> i32 system_v {
    ss0 = explicit_slot 10000
    ss1 = explicit_slot 10000

ebb0:
    v0 = iconst.i32 0
    return v0
}
```

- What do you expect to happen? What does actually happen? Does it panic, and
  if so, with which assertion?
I expect this to compile and return 0. Instead I get a linker error:
```/usr/local/bin/ld: /tmp/.tmpK629Dc: in function `main':
intcode.c:(.text.main+0xc): undefined reference to `__cranelift_probestack'```

- Which Cranelift version / commit hash / branch are you using?

Cranelift 0.49.0

- If relevant, can you include some extra information about your environment?
x86_64 Linux, Ubuntu 18.04 LTS, linking using GNU ld 2.32. I am using cranelift-faerie to emit object files.

**Comment by bjorn3** [comment]
You either need to define `__cranelift_stackprobe` or disable stack probing.

**Comment by jyn514** [comment]
What would I define `__cranelift_stackprobe` as? It looks like it's supposed to check if I overflowed the guard page: https://docs.rs/cranelift-codegen/0.51.0/cranelift_codegen/settings/struct.Flags.html#method.probestack_enabled? How would I do that?

**Comment by bjorn3** [comment]
The rust counterpart can be found at https://github.com/rust-lang/compiler-builtins/blob/master/src/probestack.rs

**Comment by jyn514** [comment]
That's a lot of inline assembly ... I think I'll just disable it for now. Thank you!

**Comment by jyn514** [comment]
BTW, why is this a function that the compiler has to provide? Shouldn't it be the same regardless of the source language?

**Comment by bjorn3** [comment]
It isn't necessarily the compiler which has to provide it, but someone has to. In case or rustc, it is `compiler-builtins`. It isn't a standardized function, so every compiler has it's own name and potentially own abi for it.

**Comment by jyn514** [comment]
Right, I was suggesting that cranelift should provide it (since it is the one emitting calls to it and knows whether or not it needs to be included in the binary).

**Comment by bjorn3** [comment]
It is impossible to write `__cranelift_probestack` in clif ir, as regalloc will probably clobber registers or allocate stack space, which is not allowed.
